File: 104.java
// 104. Maximum Depth of Binary Tree
// Given a binary tree, find its maximum depth.
//
// The maximum depth is the number of nodes along the longest path from the root node down to the farthest leaf node.
//
// Note: A leaf is a node with no children.
//
// Example:
//
// Given binary tree [3,9,20,null,null,15,7],
//     3
//    / \
//   9  20
//     /  \
//    15   7
// return its depth = 3.

// Runtime: 0 ms, faster than 100.00% of Java online submissions for Maximum Depth of Binary Tree.
// Memory Usage: 40.9 MB, less than 5.38% of Java online submissions for Maximum Depth of Binary Tree.

public class Solution {
  /**
   * Definition for a binary tree node.
   * public class TreeNode {
   *     int val;
   *     TreeNode left;
   *     TreeNode right;
   *     TreeNode(int x) { val = x; }
   * }
   */
  public int maxDepth(TreeNode root) {
    if (Objects.isNull(root)) {
      return 0;
    }
    return Math.max(checkLeaf(root.left), checkLeaf(root.right));
  }

  private int checkLeaf(TreeNode root) {
    if (Objects.isNull(root)) {
      return 1;
    }
    return 1 + Math.max(checkLeaf(root.right), checkLeaf(root.left));
  }
}

File: 104.js
// 104. Maximum Depth of Binary Tree
//
// Given the root of a binary tree, return its maximum depth.
//
// A binary tree's maximum depth is the number of nodes along the longest path from the root node down to the farthest leaf node.
//
//
//
// Example 1:
//
//
// Input: root = [3,9,20,null,null,15,7]
// Output: 3
// Example 2:
//
// Input: root = [1,null,2]
// Output: 2
//
//
// Constraints:
//
// The number of nodes in the tree is in the range [0, 104].
// -100 <= Node.val <= 100
//
//
// Runtime 114 ms Beats 5.04% of users with JavaScript
// Memory 45.46 MB Beats 35.44% of users with JavaScript
var maxDepth = function(root) {
    if (!root) return 0;
    return 1 + Math.max(maxDepth(root.left), maxDepth(root.right));  
};

// Runtime 88 ms Beats 5.63% of users with JavaScript
// Memory 45.36 MB Beats 42.81% of users with JavaScript
var maxDepth = function(root) {
    if (!root) return 0;
    let queue = [root];
    let count = 0;
    while (queue.length) {
        let size = queue.length;
        for (let i = 0; i < size; i++) {
            let cur = queue.shift();
            if (cur.left) {
                queue.push(cur.left);
            }
            if (cur.right) {
                queue.push(cur.right);
            }
        }
        count++;
    }
    return count;
};